File: frontend/src/types/note.ts
import type { PaginationParams } from "@/types/pagination";

export type Note = {
  company_id: string | null;
  contact_id: string | null;
  content: string;
  created_at: string;
  id: string;
  lead_id: string | null;
  organization_id: string;
  updated_at: string;
  user_id: string;
};

export type NoteInput = {
  company_id: string | null;
  contact_id: string | null;
  content: string;
  lead_id: string | null;
};

export type NotesListParams = PaginationParams & {
  company_id?: string;
  lead_id?: string;
  search?: string;
  sort_by?: "created_at" | "updated_at";
  sort_direction?: "asc" | "desc";
};

File: backend/app/schemas/note.py
from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, StringConstraints, field_validator

from app.schemas.common import RequestSchema

NoteContent = Annotated[str, StringConstraints(min_length=1, max_length=20_000)]


class NoteCreate(RequestSchema):
    content: NoteContent
    company_id: UUID | None = None
    contact_id: UUID | None = None
    lead_id: UUID | None = None


class NoteUpdate(RequestSchema):
    content: NoteContent | None = None
    company_id: UUID | None = None
    contact_id: UUID | None = None
    lead_id: UUID | None = None

    @field_validator("content")
    @classmethod
    def require_non_null_content(cls, value: str | None) -> str | None:
        if value is None:
            raise ValueError("Note content cannot be null")
        return value


class NoteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    user_id: UUID
    company_id: UUID | None
    contact_id: UUID | None
    lead_id: UUID | None
    content: str
    created_at: datetime
    updated_at: datetime
